kusto/tokenprovider.go

```diff
@@ -0,0 +1,50 @@
+package kusto
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+)
+
+type TokenProvider struct {
+	tokenCred   azcore.TokenCredential //Holds the received token credential as per the authorization
+	tokenScheme string                 //Contains token scheme for tokenprovider
+	customToken string                 //Holds the custom auth token to be used for authorization
+	initOnce    *sync.Once             //To ensure tokenprovider will be initialized only once while aquiring token
+	init        func()                 //Initialiser for initialising the tokencredential and cloudinfo
+	ci          CloudInfo              //Contains cloud setting meto information
+	scopes      []string               //Contains scopes of the auth token
+	err         error                  //To monitor errors in the process of init tokenprovider
+}
+
+// tokenProvider need to be received as reference, to reflect updations to the structs
+func (tkp *TokenProvider) AcquireToken(ctx context.Context) (string, string, error) {
+	if !isEmpty(tkp.customToken) {
+		return tkp.customToken, tkp.tokenScheme, nil
+	}
+
+	if tkp.tokenCred == nil {
+		// initialise the tokenCredential and cloud info
+		tkp.initOnce.Do(tkp.init)
```

Comment:
I didn't mean for you to copy-paste it to every provider, in fact in any case where you have to copy paste ask yourself if there's a better way.

Consider:

```go
tkp.initOnce.Do(func() {
tkp.init()

// Update resource URI if MFA enabled
resourceURI := tkp.ci.KustoServiceResourceID
if tkp.ci.LoginMfaRequired {
resourceURI = strings.Replace(resourceURI, ".kusto.", ".kustomfa.", 1)
}
scopes := []string{fmt.Sprintf("%s/.default", resourceURI)}
tkp.scopes = scopes
})
```